### PR TITLE
Jan 2020 milestone 2️⃣0️⃣2️⃣0️⃣

### PR DESCRIPTION
**Pull request by dhmlau**
# Jan milestone

## Overview

- Take some time to:
  - fix bugs, improve developer experience, fix docs
  - triage open issues across all the repos, not just `loopback-next`
- Migration work continues to be one of the focus
- Focus on issues/PRs for loopback-connector-mongodb

## Goals

- Migration guide between LB3 and LB4 MVP #453

  - [ ] How to migrate boot scripts #3957

- Improve developer experience

  - [ ] Expose isActive() method in Transaction interface #3471
  - [ ] Missing model.ts after running lb4 openapi command #4135
  - [ ] Model strict: false is ignored with Postgre SQL repository #4042

- Bugs

  - [ ] :running: Docs: Improve TodoList Tutorial instructions and completed tutorial
        'lb4 example todo-list' #4161
  - [x] lb3 - Relations on uuid
        https://github.com/strongloop/loopback/issues/4287

- Create models & REST APIs dynamically at runtime

  - [ ] [Needs estimates] Dynamic binding/rebinding of controllers after app
        start #433

- Update dependencies

  - [ ] [Needs estimates] chore: update dependency change-case to v4 #4233
  - [ ] [Needs estimates] update dependency ts-morph to v6 #4274

- [ ] [spike] API Connect / LoopBack 4 integration #4115

- _Miscellaneous_
  - [ ] Spike: Migration guide from LB3 - Authentication & authorization
#3719
  - [ ] :running: API Explorer ignores filter options where, fields, and order #2208
  - [ ] lb4 repository,model,controller commands generate artifacts with 'lint'
issues #4205

## Stretch goals

- Migration guide between LB3 and LB4 MVP #453

  - [ ] How to migrate user-defined model methods #3949
  - [ ]  How to migrate remoting hooks #3950

- Create models & REST APIs dynamically at runtime

  - [ ] How to build models, repositories and controllers dynamically at runtime
        #4296 [needs estimate]

- Bugs

  - [ ] [community contribution?] Preserve custom type of auto-generated id
    property #3602 [needs estimate]
  - [ ] Column names in lowercase using lb4 discover method -
        SQL connector #3343
  - [ ] :running: lb4 relation should discard changes when something failed #3287 

- PR reviews
  - [ ] feat(extension-logging): add integration with winston and fluentd
        logging, https://github.com/strongloop/loopback-next/pull/4117
  - [ ] [RFC] feat: add base code for the image of Appsody LoopBack stack,
        https://github.com/strongloop/loopback-next/pull/4250 [draft pr still in discussion]

### CODE REVIEW

File: docs/_jan-milestone_.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  - [ ] [3] Missing model.ts after running lb4 openapi command #4135
+  - [ ] [2] Model strict: false is ignored with Postgre SQL repository #4042
+  - [ ] [needs estimates] Column names in lowercase using lb4 discover method -
+        SQL connector #3343
+
+- [ ] Bugs
+
+  - [ ] [3] Docs: Improve TodoList Tutorial instructions and completed tutorial
+        'lb4 example todo-list' #4161
+
```
Comment (bajtos) [inline docs/_jan-milestone_.md]: I am proposing to add the following bug(s) to the Jan milestone:

- https://github.com/strongloop/loopback/issues/4287 lb3 - Relations on uuid

Comment (bajtos) [reply docs/_jan-milestone_.md]: - Preserve custom type of auto-generated id property #3602

File: docs/_jan-milestone_.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  - [ ] [5] How to migrate remoting hooks #3950
+
+- [ ] From model definition to REST API with no custom repository/controller
+      classes #2036
+
+  - [ ] [3]Add CrudRestApiBuilder to @loopback/rest-crud #3737
+  - [ ] [3]Example app showing CrudRestApiBuilder #3738
+
+- [ ] [3][spike] API Connect / LoopBack 4 integration #4115
+
```
Comment (bajtos) [inline docs/_jan-milestone_.md]: This dependency upgrade requires deeper investigation, let's plan some time in January to get it done.

- chore: update dependency change-case to v4 #4233

File: docs/_jan-milestone_.md (hunk 1 above)
Comment (dhmlau) [reply docs/_jan-milestone_.md]: for #3602, will put it in the stretch goal. It seems like @frbuceta will pick it up.

**Comment by dhmlau** [comment]
@bajtos, I've incorporated your comments and also moved some of the migration guide tasks to stretch goal. We currently are committing 26 points + 2 tasks need to be estimated.  It would be a good committed number of story points due to vacation time and we can have more flexibility to fix bugs/usability issues.

**Comment by dhmlau** [comment]
@bajtos, I've updated the Jan milestone based on our discussion today:
- have more focus / people to work on the migration story 
- remove the "from model definition to REST" epic / tasks. 

Could you please add the migration tasks that you think we should work on in Jan? Thanks.

File: docs/_jan-milestone_.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- [ ] Bugs
+
+  - [ ] [3] Docs: Improve TodoList Tutorial instructions and completed tutorial
+        'lb4 example todo-list' #4161
+  - [ ] [Needs Estimates] lb3 - Relations on uuid
+        https://github.com/strongloop/loopback/issues/4287
+
+- [ ] [3][spike] API Connect / LoopBack 4 integration #4115
+
+- [ ] [Needs estimates] chore: update dependency change-case to v4 #4233
```
Comment (dhmlau) [inline docs/_jan-milestone_.md]: similarly, we can add chore: update dependency ts-morph to v6 https://github.com/strongloop/loopback-next/pull/4274?

File: docs/_jan-milestone_.md (hunk 2 above)
Comment (bajtos) [reply docs/_jan-milestone_.md]: Another similar one:

- chore: update dependency ts-morph to v6 #4274

File: docs/_jan-milestone_.md (hunk 3 above)
Comment (bajtos) [reply docs/_jan-milestone_.md]: Yes please!

File: docs/_jan-milestone_.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+## Stretch goals
+
+- [ ] Spike: migrate tests to Jest (ts-jest) #3159
```
Comment (bajtos) [inline docs/_jan-milestone_.md]: Let's remove this item. Based on my discussion with @boneskull, we would like to try few experiments with Mocha to see if we can get DX that's good enough.

File: docs/_jan-milestone_.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  - [ ] [3] How to migrate datasources from LB3 to LB4 #3946
+  - [ ] [3] How to migrate boot scripts #3957
+
+- [ ] Improve developer experience
+
+  - [ ] [3] Expose isActive() method in Transaction interface #3471
+  - [ ] [3] Missing model.ts after running lb4 openapi command #4135
+  - [ ] [2] Model strict: false is ignored with Postgre SQL repository #4042
+
```
Comment (bajtos) [inline docs/_jan-milestone_.md]: IIRC, we wanted to also add one or two tasks for defining models at runtime, see https://github.com/strongloop/loopback-next/pull/4116/files#r357149740

```markdown
#### Create models & REST APIs dynamically at runtime

- Dynamic binding/rebinding of controllers after app start #433
- How to build models, repositories and controllers dynamically at runtime #4296
```

I am proposing to commit the first task for January and the second for February. Thoughts?

Comment (dhmlau) [reply docs/_jan-milestone_.md]: I've put #433 to committed stories in Jan, and #4296 as stretch goal.

**Comment by dhmlau** [comment]
https://github.com/strongloop/loopback-next/issues/4376 is opened. closing this PR. Thanks!

**Pull request closed without merging**